unit_testing/LightController/LightController_tests.cpp

```diff
@@ -0,0 +1,432 @@
+       protocol.*/
+
+    // Invalid port   
+    writeLightControllerMessage("p4 FFFFFF\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p4 FFFFFF"
+    ));
+
+    writeLightControllerMessage("p0 FFFFFF\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p0 FFFFFF"
+    ));
+
+    // Wrong hex code (too short)
+    writeLightControllerMessage("p1 13908\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p1 13908"
+    ));
+
+    // Wrong hex code (too long)
+    writeLightControllerMessage("p1 13908CC\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p1 13908CC"
+    ));
+
+    // Wrong hex code (non valid hex characters)
+    writeLightControllerMessage("p1 ACACFM\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p1 ACACFM"
+    ));
+}
+
+TEST_F(LightControllerTest, LEDSwitching)
+{
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+    LightControllerLoop(bootMsg);
+
+    /* --- Turn on p1 minimum amount of time for each RGB component --- */
+    writeLightControllerMessage("p1 010101\n");
+    LightControllerLoop(bootMsg);
+
+    // LED switching interrupt 
+    goToTick(1);
+
+    // First three GPIOs should be on and the rest should be off
+    for(int i = 0; i < 3; i++) EXPECT_TRUE(stmGetGpio(*ChCtrl[i]));
+    for(int i = 3; i < LED_CHANNELS*NO_COLORS; i++) EXPECT_FALSE(stmGetGpio(*ChCtrl[i]));
+
+    goToTick(2);
+    for(int i = 0; i < LED_CHANNELS*NO_COLORS; i++) EXPECT_FALSE(stmGetGpio(*ChCtrl[i]));
+
+    // Reset counter
+    goToTick(0);
+
+    // Check that it keeps switching correctly after overflow
+    goToTick(1);
+
+    // First three GPIOs should be on and the rest should be off
+    for(int i = 0; i < 3; i++) EXPECT_TRUE(stmGetGpio(*ChCtrl[i]));
+    for(int i = 3; i < LED_CHANNELS*NO_COLORS; i++) EXPECT_FALSE(stmGetGpio(*ChCtrl[i]));
+
+
+    /* --- Turn on p1 white LED all time --- */
+    goToTick(0);
+    writeLightControllerMessage("p1 FFFFFF\n");
+    LightControllerLoop(bootMsg);
+
+    // LED switching interrupt 
+    goToTick(100);
+    for(int i = 0; i < LED_CHANNELS*NO_COLORS; i++){
+        if (i == 3){
+            EXPECT_TRUE(stmGetGpio(*ChCtrl[i]));
+            continue;
+        }
+        EXPECT_FALSE(stmGetGpio(*ChCtrl[i]));
+    } 
+
+
+    /* --- Turn on p2 RGB --- */
+    goToTick(0);
+    writeLightControllerMessage("p1 000000\n");
+
+    // 0xAA = 170
+    writeLightControllerMessage("p2 AAAAAA\n");
+    LightControllerLoop(bootMsg);
+    goToTick(170);
+    for(int i = 0; i < LED_CHANNELS*NO_COLORS; i++){
+        if (i >= 4 && i < 7){
+            EXPECT_TRUE(stmGetGpio(*ChCtrl[i]));
+            continue;
+        }
+        EXPECT_FALSE(stmGetGpio(*ChCtrl[i]));
+    } 
+
+    goToTick(171);
+    for(int i = 0; i < LED_CHANNELS*NO_COLORS; i++) EXPECT_FALSE(stmGetGpio(*ChCtrl[i]));
+
+
+    /* --- Turn on p3 white while p2 is running --- */
+    goToTick(0);
```

Comment:
It might be more interesting to try this while still at tick 170/171, or some other number other than 0